opt: add cost for table descriptor fetch during virtual scan

rytaft opened the pull request:
This commit adds a cost equal to `10*randIOCostFactor` for each virtual
scan in order to represent the cost of fetching table descriptors.
This cost is especially important when perfoming lookup joins, because
the descriptors may need to be fetched on each lookup. As a result of
this change, the optimizer is much less likely to plan a lookup join
into a virtual table.

Fixes #55140

Release note (performance improvement): Adjusted the cost model in
the optimizer so that the optimizer is less likely to plan a lookup
join into a virtual table. Performing a lookup join into a virtual
table is expensive, so this change will generally result in better
performance for queries involving joins with virtual tables.

rytaft committed:
opt: add cost for table descriptor fetch during virtual scan

This commit adds a cost equal to 10*randIOCostFactor for each virtual
scan in order to represent the cost of fetching table descriptors.
This cost is especially important when perfoming lookup joins, because
the descriptors may need to be fetched on each lookup. As a result of
this change, the optimizer is much less likely to plan a lookup join
into a virtual table.

This commit also includes some fixes to the test catalog to provide
better support for testing virtual tables with indexes.

Fixes #55140

Release note (performance improvement): Adjusted the cost model in
the optimizer so that the optimizer is less likely to plan a lookup
join into a virtual table. Performing a lookup join into a virtual
table is expensive, so this change will generally result in better
performance for queries involving joins with virtual tables.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/55833)
<!-- Reviewable:end -->

rytaft commented:
Hold off on reviewing this -- it doesn't seem to work yet.





<!-- Sent from Reviewable.io -->

rafiss commented on pkg/sql/opt/xform/testdata/coster/virtual-scan:
```diff
@@ -4,11 +4,11 @@ SELECT * FROM information_schema.schemata WHERE SCHEMA_NAME='public'
 select
```
i think we'd also want a test (not sure if here or somewhere else) that verifies that we do use the virtual lookup join when the join condition only selects one row from the virtual index.

rytaft committed:
sql,opt: export pg_catalog table schemas for use by optimizer

This commit moves the schema definitions of all tables in pg_catalog to
a new file in sql/vtable called pg_catalog.go and exports them so they
can be used by the optimizer and elsewhere. This is similar to what was
already done for the information_schema tables.

Release note: None

rytaft reviewed (commented):
Went down a deep rabbit hole of trying to improve our support for testing virtual tables in the optimizer, so I added a commit before the one affecting the coster.

There's still lots more to fix with general optimizer support for virtual tables, particularly related to collecting stats (we don't have any stats right now). But hopefully this costing change will be an improvement for now.

PTAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55833#-:-MKD5sHX8wg7Kc1iv4yr:b1xsbio)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @RaduBerinde, and @rafiss)

___
*[pkg/sql/opt/xform/testdata/coster/virtual-scan, line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55833#-MKC-3lCq3dHtkBTmQUQ-r1-4:-MKD5iwP41-6-fbljH8V:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b401b0a8c35b558c9d271c23fecb679813a4e5e/pkg/sql/opt/xform/testdata/coster/virtual-scan#L4)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think we'd also want a test (not sure if here or somewhere else) that verifies that we do use the virtual lookup join when the join condition only selects one row from the virtual index.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
lgtm from my end. thank you!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55833#-:-MKHPjfF49d7gMz1E4Fp:b-mtwx1m)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @RaduBerinde, and @rafiss)



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55833#-:-MKHQeo4A5B8O6wpErXL:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @RaduBerinde, and @rafiss)



<!-- Sent from Reviewable.io -->

rytaft commented:
TFTRs!

bors r+





<!-- Sent from Reviewable.io -->

rytaft commented:
Fixed a test failure due to re-ordering the output rows. Added `rowsort` to avoid these failures in the future. 





<!-- Sent from Reviewable.io -->

rytaft commented:
bors r+





<!-- Sent from Reviewable.io -->

rytaft commented:
Flake https://github.com/cockroachdb/cockroach/issues/50032

bors r+





<!-- Sent from Reviewable.io -->

rytaft commented:
Same problem

bors r+





<!-- Sent from Reviewable.io -->

rytaft commented:
Trying again since I think that issue has been fixed now.

bors r+





<!-- Sent from Reviewable.io -->
